nexus/db-queries/src/db/datastore/instance.rs

```diff
@@ -1188,28 +1167,108 @@ impl DataStore {
+                    // Not updated, because the update is no change..
+                    return Ok(());
+                }
+
+                if !InstanceState::NOT_INCARNATED_STATES
+                    .contains(&r.found.runtime().nexus_state)
+                {
+                    return Err(err.bail(Error::conflict(
+                        "instance must be stopped to set boot disk",
+                    )));
+                }
+
+                // There should be no other reason the update fails on an
+                // existing instance.
+                return Err(err.bail(Error::internal_error(
+                    "unable to reconfigure instance boot disk",
+                )));
+            }
+            UpdateStatus::Updated => Ok(()),
+        }
+    }
+
+    /// Set an instance's CPU count and memory size to the provided values,
+    /// within an existing transaction.
+    ///
+    /// Does not allow setting sizes of running instances to ensure that if an
+    /// instance is running, its resource reservation matches what we record in
+    /// the database.
+    async fn instance_set_size_on_conn(
+        &self,
+        conn: &async_bb8_diesel::Connection<DbConnection>,
+        err: &OptionalError<Error>,
+        authz_instance: &authz::Instance,
+        ncpus: InstanceCpuCount,
+        memory: ByteCount,
+    ) -> Result<(), diesel::result::Error> {
+        use db::schema::instance::dsl as instance_dsl;
+
+        let r = diesel::update(instance_dsl::instance)
+            .filter(instance_dsl::id.eq(authz_instance.id()))
+            .filter(
+                instance_dsl::state
+                    .eq_any(InstanceState::NOT_INCARNATED_STATES),
+            )
+            .filter(
+                instance_dsl::ncpus
+                    .ne(ncpus)
+                    .or(instance_dsl::memory.ne(memory)),
+            )
+            .set((
+                instance_dsl::ncpus.eq(ncpus),
+                instance_dsl::memory.eq(memory),
+            ))
+            .check_if_exists::<Instance>(authz_instance.id())
+            .execute_and_check(&conn)
+            .await?;
+        match r.status {
+            UpdateStatus::NotUpdatedButExists => {
+                if (r.found.ncpus, r.found.memory) == (ncpus, memory) {
+                    // Not updated, because the update is no change..
+                    return Ok(());
+                }
+
+                if !InstanceState::NOT_INCARNATED_STATES
+                    .contains(&r.found.runtime().nexus_state)
+                {
+                    return Err(err.bail(Error::conflict(
+                        "instance must be stopped to be resized",
+                    )));
+                }
+
+                // There should be no other reason the update fails on an
+                // existing instance.
+                return Err(err.bail(Error::internal_error(
+                    "unable to reconfigure instance size",
+                )));
```

Comment:
similarly, this indicates some kind of bug in nexus, so it might be nice to record additional information about it, in case we end up debugging this in the future? not a big deal though.

Reply:
same as above, went for a bit less than the whole instance record but hopefully enough to help us debug later [here](https://github.com/oxidecomputer/omicron/pull/6774/commits/422ae418f407a558ef6c18cbae45c9ff79fa9421#diff-a234811eb5ee38fdcf626cd6514deb7dbfc8b9e2067cfa3b4a6f111d72a07e53R1292-R1298). ty!